config/config.go

```diff
@@ -364,7 +373,22 @@ func WatchConfig(ctx context.Context) error {
 	for {
 		md, eTag, webError = getMetadata(fmt.Sprintf("?recursive=true&alt=json&wait_for_change=true&last_etag=%s&timeout_sec=%d", lEtag.get(), osConfigMetadataPollTimeout))
 		if webError == nil && eTag != lEtag.get() {
-			break
+			lEtag.set(eTag)
+			var metadataConfig metadataJSON
+			if err := json.Unmarshal(md, &metadataConfig); err != nil {
+				return err
+			}
+
+			newAgentConfig := createConfigFromMetadata(metadataConfig)
+			newAgentConfigHash := newAgentConfig.asSha256()
+
+			if lAgentConfigHash != newAgentConfigHash {
+				agentConfigMx.Lock()
```

Comment:
Since `agentConfig` and `lAgentConfigHash` represent the same data, should they both be protected by the lock? Or would it be better to calculate `lAgentConfigHash` on the fly instead of caching it and dealing with redundant state?

Reply:
i thought it would be unnecessary to use lock for `lAgentConfigHash` because it is not a shared variable.
Also, `agentConfig` is written only by the current thread. So, the operations, calculating hash of `agentConfig` and updating `agentConfig`, are sequential.

Reply:
We don't need to worry about trying to be super efficient with the lock, just do this for simplicity and to avoid more global vars:

```
agentConfigMx.Lock()
if agentConfig.asSha256() != newAgentConfig.asSha256() {
  agentConfig = newAgentConfig
  agentConfigMx.Unlock()
  break
}
agentConfigMx.Unlock()
```

Reply:
sure

Reply:
done